File: python/data_structures/tree/avl/avl.py
"""AVL 树：插入与删除。"""

from __future__ import annotations


class AVLNode:
    __slots__ = ("key", "left", "right", "height")

    def __init__(self, key: int) -> None:
        self.key = key
        self.left: AVLNode | None = None
        self.right: AVLNode | None = None
        self.height = 1


def _h(n: AVLNode | None) -> int:
    return n.height if n is not None else 0


def _upd(n: AVLNode) -> None:
    n.height = 1 + max(_h(n.left), _h(n.right))


def _bf(n: AVLNode) -> int:
    return _h(n.left) - _h(n.right)


def _rotate_right(y: AVLNode) -> AVLNode:
    x = y.left
    assert x is not None
    t2 = x.right
    x.right = y
    y.left = t2
    _upd(y)
    _upd(x)
    return x


def _rotate_left(x: AVLNode) -> AVLNode:
    y = x.right
    assert y is not None
    t2 = y.left
    y.left = x
    x.right = t2
    _upd(x)
    _upd(y)
    return y


class AVLTree:
    def __init__(self) -> None:
        self.root: AVLNode | None = None

    def insert(self, key: int) -> None:
        self.root = self._insert(self.root, key)

    def _insert(self, n: AVLNode | None, key: int) -> AVLNode:
        if n is None:
            return AVLNode(key)
        if key < n.key:
            n.left = self._insert(n.left, key)
        elif key > n.key:
            n.right = self._insert(n.right, key)
        else:
            return n
        _upd(n)
        b = _bf(n)
        if b > 1 and n.left is not None and key < n.left.key:
            return _rotate_right(n)
        if b > 1 and n.left is not None and key > n.left.key:
            n.left = _rotate_left(n.left)
            return _rotate_right(n)
        if b < -1 and n.right is not None and key > n.right.key:
            return _rotate_left(n)
        if b < -1 and n.right is not None and key < n.right.key:
            n.right = _rotate_right(n.right)
            return _rotate_left(n)
        return n

    def delete(self, key: int) -> None:
        self.root = self._delete(self.root, key)

    def _min_node(self, n: AVLNode) -> AVLNode:
        while n.left is not None:
            n = n.left
        return n

    def _delete(self, n: AVLNode | None, key: int) -> AVLNode | None:
        if n is None:
            raise KeyError(key)
        if key < n.key:
            n.left = self._delete(n.left, key)
        elif key > n.key:
            n.right = self._delete(n.right, key)
        else:
            if n.left is None:
                return n.right
            if n.right is None:
                return n.left
            succ = self._min_node(n.right)
            n.key = succ.key
            n.right = self._delete(n.right, succ.key)
        return self._rebalance(n)

    def _rebalance(self, n: AVLNode | None) -> AVLNode | None:
        if n is None:
            return None
        _upd(n)
        b = _bf(n)
        if b > 1 and n.left is not None and _bf(n.left) >= 0:
            return _rotate_right(n)
        if b > 1 and n.left is not None and _bf(n.left) < 0:
            n.left = _rotate_left(n.left)
            return _rotate_right(n)
        if b < -1 and n.right is not None and _bf(n.right) <= 0:
            return _rotate_left(n)
        if b < -1 and n.right is not None and _bf(n.right) > 0:
            n.right = _rotate_right(n.right)
            return _rotate_left(n)
        return n

    def inorder(self) -> list[int]:
        out: list[int] = []

        def dfs(x: AVLNode | None) -> None:
            if x is None:
                return
            dfs(x.left)
            out.append(x.key)
            dfs(x.right)

        dfs(self.root)
        return out


if __name__ == "__main__":
    t = AVLTree()
    for x in [10, 20, 30, 40, 50, 25]:
        t.insert(x)
    t.delete(20)
    ord_ = t.inorder()
    assert ord_ == sorted(ord_)
    t.insert(10)
    assert t.inorder().count(10) == 1
    try:
        t.delete(999)
        raise AssertionError("expected KeyError")
    except KeyError:
        pass
    print("AVL OK", ord_)

File: cpp/data_structures/tree/avl/avl.cpp
#include <alg_std.hpp>
#include <cassert>
using namespace std;

struct Node {
    int key, h;
    Node *l, *r;
    explicit Node(int k) : key(k), h(1), l(nullptr), r(nullptr) {}
};

int height(Node* n) { return n ? n->h : 0; }

void upd(Node* n) { n->h = 1 + max(height(n->l), height(n->r)); }

int bf(Node* n) { return height(n->l) - height(n->r); }

Node* rotR(Node* y) {
    Node* x = y->l;
    Node* t2 = x->r;
    x->r = y;
    y->l = t2;
    upd(y);
    upd(x);
    return x;
}

Node* rotL(Node* x) {
    Node* y = x->r;
    Node* t2 = y->l;
    y->l = x;
    x->r = t2;
    upd(x);
    upd(y);
    return y;
}

Node* insert(Node* n, int key) {
    if (!n) return new Node(key);
    if (key < n->key)
        n->l = insert(n->l, key);
    else if (key > n->key)
        n->r = insert(n->r, key);
    else
        return n;
    upd(n);
    int b = bf(n);
    if (b > 1 && n->l && key < n->l->key) return rotR(n);
    if (b > 1 && n->l && key > n->l->key) {
        n->l = rotL(n->l);
        return rotR(n);
    }
    if (b < -1 && n->r && key > n->r->key) return rotL(n);
    if (b < -1 && n->r && key < n->r->key) {
        n->r = rotR(n->r);
        return rotL(n);
    }
    return n;
}

Node* minNode(Node* n) {
    while (n->l) n = n->l;
    return n;
}

Node* rebalance(Node* n);

Node* del(Node* n, int key) {
    if (!n) throw out_of_range("del");
    if (key < n->key)
        n->l = del(n->l, key);
    else if (key > n->key)
        n->r = del(n->r, key);
    else {
        if (!n->l || !n->r) {
            Node* t = n->l ? n->l : n->r;
            delete n;
            return t;
        }
        Node* s = minNode(n->r);
        n->key = s->key;
        n->r = del(n->r, s->key);
    }
    return rebalance(n);
}

Node* rebalance(Node* n) {
    if (!n) return nullptr;
    upd(n);
    int b = bf(n);
    if (b > 1 && n->l && bf(n->l) >= 0) return rotR(n);
    if (b > 1 && n->l && bf(n->l) < 0) {
        n->l = rotL(n->l);
        return rotR(n);
    }
    if (b < -1 && n->r && bf(n->r) <= 0) return rotL(n);
    if (b < -1 && n->r && bf(n->r) > 0) {
        n->r = rotR(n->r);
        return rotL(n);
    }
    return n;
}

void inorder(Node* n, vector<int>& out) {
    if (!n) return;
    inorder(n->l, out);
    out.push_back(n->key);
    inorder(n->r, out);
}

int main() {
    Node* root = nullptr;
    for (int x : {10, 20, 30, 40, 50, 25}) root = insert(root, x);
    root = del(root, 20);
    vector<int> ord;
    inorder(root, ord);
    vector<int> cp = ord;
    sort(cp.begin(), cp.end());
    assert(ord == cp);
    cout << "AVL OK" << endl;
    return 0;
}
